### PR TITLE
drivers/sx127x: Add initial support for multi interrupt pins

### PR DESCRIPTION
**Pull request by leandrolanzieri**
### Contribution description
[Some boards](https://learn.watterott.com/bee-modules/lora/) that breakout the sx127x radios unify the interrupt pins into a single one. This is an initial support for those cases. When the *multi dio* pin triggers an interrupt the registers on the sx127x are checked to identify the interrupt cause, and the appropriate isr is executed. 

Currently this fix is needed by the 'SenseBox with SAMD21' board.

### Usage
To enable the "multi dio" pin the macro `SX127X_USE_DIO_MULTI` must be defined. This can be done by adding `CFLAGS += -DSX127X_USE_DIO_MULTI` to the makefile of the board.

### CODE REVIEW

File: boards/sensebox_samd21/include/board.h [hunk 1]
```diff
@@ -123,24 +123,33 @@ extern "C" {
...
-                                            .paselect  = SX127X_PARAM_PASELECT }
+#define SX127X_PARAM_DIO2           GPIO_UNDEF
+
+#define SX127X_PARAM_DIO3           GPIO_UNDEF
+
+#define SX127X_PARAM_DIOMULTI       XBEE1_INT_PIN       /* D24 */
+
+#define SX127X_PARAM_PASELECT       (SX127X_PA_RFO)
+
+#define SX127X_PARAMS               { .spi       = SX127X_PARAM_SPI,           \
```
Comment (aabadie) [inline boards/sensebox_samd21/include/board.h]: This define is not needed since all `SX127X_PARAMS_*` are already defined and will be used in the `sx127x_params.h` file.

File: drivers/sx127x/include/sx127x_params.h [hunk 2]
```diff
@@ -33,46 +33,62 @@ extern "C" {
  * @{
  */
 #ifndef SX127X_PARAM_SPI
-#define SX127X_PARAM_SPI                    (SPI_DEV(0))
+#define SX127X_PARAM_SPI          (SPI_DEV(0))
```
Comment (aabadie) [inline drivers/sx127x/include/sx127x_params.h]: unrelated alignment change, please revert (here and below)

File: drivers/include/sx127x.h [hunk 3]
```diff
@@ -208,6 +211,9 @@ typedef struct {
     gpio_t dio3_pin;                   /**< Interrupt line DIO3 (CAD done) */
     gpio_t dio4_pin;                   /**< Interrupt line DIO4 (not used) */
     gpio_t dio5_pin;                   /**< Interrupt line DIO5 (not used) */
+#ifdef SX127X_USE_DIO_MULTI
+    gpio_t dio_multi_pin;              /**< Interrupt line for muliple IRQs */
```
Comment (aabadie) [inline drivers/include/sx127x.h]: s/muliple/multiple/

File: drivers/sx127x/include/sx127x_params.h [hunk 4]
```diff
@@ -33,46 +33,62 @@ extern "C" {
...
-#define SX127X_PARAMS                       { .spi       = SX127X_PARAM_SPI,     \
-                                              .nss_pin   = SX127X_PARAM_SPI_NSS, \
-                                              .reset_pin = SX127X_PARAM_RESET,   \
-                                              .dio0_pin  = SX127X_PARAM_DIO0,    \
-                                              .dio1_pin  = SX127X_PARAM_DIO1,    \
-                                              .dio2_pin  = SX127X_PARAM_DIO2,    \
-                                              .dio3_pin  = SX127X_PARAM_DIO3,    \
-                                              .paselect  = SX127X_PARAM_PASELECT }
+#ifdef SX127X_USE_DIO_MULTI
+    #define SX127X_PARAMS             { .spi       = SX127X_PARAM_SPI,          \
```
Comment (aabadie) [inline drivers/sx127x/include/sx127x_params.h]: no need to indent this `#define`, same below

File: drivers/sx127x/sx127x.c [hunk 5]
```diff
@@ -207,35 +219,81 @@ static void sx127x_on_dio3_isr(void *arg)
...
+    if (dev->params.dio0_pin != GPIO_UNDEF) {
+        res = gpio_init_int(dev->params.dio0_pin, GPIO_IN, GPIO_RISING,
+                                sx127x_on_dio0_isr, dev);
+        if (res < 0) {
+            DEBUG("[sx127x] error: failed to initialize DIO0 pin\n");
+            return res;
+        }
+    }
+    else {
+        DEBUG("[sx127x] error: no DIO0 pin defined\n");
```
Comment (aabadie) [inline drivers/sx127x/sx127x.c]: is it still an error ?

Comment (aabadie) [reply drivers/sx127x/sx127x.c]: I would do this differently:
```c
#ifndef SX127X_USE_DIO_MULTI
   /* Check if DIO0 pin is defined */
    if (dev->params.dio0_pin != GPIO_UNDEF) {
        res = gpio_init_int(dev->params.dio0_pin, GPIO_IN, GPIO_RISING,
                            sx127x_on_dio0_isr, dev);
        if (res < 0) {
            DEBUG("[sx127x] error: failed to initialize DIO0 pin\n");
            return res;
        }
    }

    /* check if DIO1 pin is defined */
    if (dev->params.dio1_pin != GPIO_UNDEF) {
        res = gpio_init_int(dev->params.dio1_pin, GPIO_IN, GPIO_RISING,
                            sx127x_on_dio1_isr, dev);
        if (res < 0) {
            DEBUG("[sx127x] error: failed to initialize DIO1 pin\n");
            return res;
        }
    }
#else
/* your new multi dio stuff here */
#endif
```

File: drivers/sx127x/sx127x_netdev.c [hunk 6]
```diff
@@ -240,6 +240,36 @@ static void _isr(netdev_t *netdev)
     sx127x_t *dev = (sx127x_t *) netdev;
 
     uint8_t irq = dev->irq;
+
+    #ifdef SX127X_USE_DIO_MULTI
```
Comment (aabadie) [inline drivers/sx127x/sx127x_netdev.c]: No need to indent this `#ifdef`

File: drivers/sx127x/sx127x.c (hunk 5 above)
Comment (leandrolanzieri) [reply drivers/sx127x/sx127x.c]: I did it that way because with this radio I think you at least need the DIO0 defined. So in the case that DIO0 is not defined and you are not using the "multi" mode it should fail. If you are using the "multi" mode and no DIO0 is defined, it would check if the MULTI_DIO pin is defined.

Comment (leandrolanzieri) [reply drivers/sx127x/sx127x.c]: @aabadie Did you have a chance to look at this?

Comment (aabadie) [reply drivers/sx127x/sx127x.c]: > Did you have a chance to look at this?

Yes, and I still want this to be changed since it can be done simpler. I edited my code snippet above with what I have in mind.
I also think the `sx127x_on_dioX_isr` static functions could be excluded if `SX127X_USE_DIO_MULTI` is defined. Can you check if that save a few bytes ?

Comment (leandrolanzieri) [reply drivers/sx127x/sx127x.c]: Modified so it would only initialize the Multi dio pin if `SX127X_USE_DIO_MULTI` is defined. Also excluded the `sx127x_on_diox_isr` functions.

**Comment by leandrolanzieri** [comment]
@aabadie ping?

**Comment by leandrolanzieri** [comment]
Rebased and squashed.

**Comment by aabadie** [comment]
Thanks @leandrolanzieri !
